### PR TITLE
ROSTESTS-263. Check/Wait for start/stop of services to complete

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

Detect service stop failures.
And start failures too, if such happen.

JIRA issue: [ROSTESTS-263](https://jira.reactos.org/browse/ROSTESTS-263)

## TODO

Simple tests passed/failed on WXP, as intented.
`rosautotest` passed on ROS r76032.

- [X] Run `rosautotest` on Windows S03. // Passed. Thanks to @Doug-Lyons!
- [X] Run `rosautotest` on recent ReactOS. // Skipped: too complicated to do on my VPC 2004 due to recent regressions. I don't expect any difference anyway.

### CODE REVIEW

**Comment by SergeGautherie** [comment]
On VPC 2004, partial tests:
- WXP: RUNNING + STOPPED. (Or STOP_PENDING when testing targeted issues.)
- ROS: RUNNING + (RUNNING then STOPPED).

WXP behaves as if its `ControlService( SERVICE_CONTROL_STOP )` includes `Sleep(0)`...
Maybe ROS should do that too, if WS03 behaves like that too?

File: modules/rostests/kmtests/kmtest/service.c [hunk 1]
```diff
@@ -227,6 +246,28 @@ KmtStopService(
...
+
+    if (ServiceStatus.dwCurrentState == SERVICE_STOPPED)
+    {
+        trace("Service %ls state = STOPPED, immediately\n", ServiceNameOut);
+        goto cleanup;
+    }
+
+    // Give up timeslice, so the service has a chance to actually stop.
+    // If this is not enough, then probably increase to 1000 (= 1 second)...
+    Sleep(0);
```
Comment (ThFabba) [inline modules/rostests/kmtests/kmtest/service.c]: This is just asking for a race condition. You need a wait loop for it to be reliable.

**Comment by ThFabba** [comment]
> WXP behaves as if its ControlService( SERVICE_CONTROL_STOP ) includes Sleep(0)...
Maybe ROS should do that too, if WS03 behaves like that too?

That's not a useful test, it's entirely dependent on (a) the service's code and (b) scheduling behavior, both of which cannot be relied on on any version of the OS.

File: modules/rostests/kmtests/kmtest/service.c (hunk 1 above)
Comment (SergeGautherie) [reply modules/rostests/kmtests/kmtest/service.c]: Done.

**Comment by SergeGautherie** [comment]
Not WIP anymore.
Please, can someone test this on WS03, as need be...

**Comment by SergeGautherie** [comment]
Minor comment fix.

**Comment by SergeGautherie** [comment]
Not WIP anymore.

File: modules/rostests/kmtests/kmtest/service.c [hunk 2]
```diff
@@ -100,6 +102,103 @@ KmtCreateService(
...
+DWORD
+KmtCheckServiceState(
+    IN PCWSTR ServiceName OPTIONAL,
+    IN SC_HANDLE ServiceHandle,
+    IN DWORD ExpectedServiceState)
+{
+    DWORD Error = ERROR_SUCCESS;
+    SERVICE_STATUS ServiceStatus;
+    DWORD StartTime = GetTickCount();
+    DWORD Timeout = 30000; // 30 second timeout
```
Comment (ThFabba) [inline modules/rostests/kmtests/kmtest/service.c]: Comment adds no information. Unless you consider it non-obvious that these are milliseconds, in which case `#define SECOND 1000` and let the code speak for itself.

Comment (SergeGautherie) [reply modules/rostests/kmtests/kmtest/service.c]: Replaced with in-between solution: `30 * 1000`.

NB:
30-seconds feels pretty long to wait for, when it happens (for broken tests).
Tell me whether you want/agree to reduce it to 10 for example.

Comment (ThFabba) [reply modules/rostests/kmtests/kmtest/service.c]: Good point. I agree 10 seconds seems more than enough.

Comment (SergeGautherie) [reply modules/rostests/kmtests/kmtest/service.c]: Done. With `Sleep(1 * 1000);` too.

**Comment by SergeGautherie** [comment]
Updated a header email address.

**Comment by SergeGautherie** [comment]
Any reason why this PR remains `WIP`?

**Comment by learn-more** [comment]
Probably this is the reason?

**Comment by SergeGautherie** [comment]
> Probably this is the reason?

I did some basic tests on WXP and do not expect any difference on WS03.
I am interested in results mostly wrt ROSTESTS-266.
I asked Doug, who did not respond, and Stas', who has no WS03 atm.
Then, who/how can this be done?

**Comment by gedmurphy** [comment]
I don't really get the point of this change. The StartService and ControlService only return success if the service reached the requested state.
IMO, this PR should be closed.

**Comment by SergeGautherie** [comment]
> The StartService and ControlService only return success if the service reached the requested state.

This statement is false, at least for `ControlService(SERVICE_CONTROL_STOP)`! Success means only that the command was sent.
What makes you think otherwise?

**Comment by gedmurphy** [comment]
Yep sorry, ignore my last comment, I got that wrong, 
I was in the pub and browsing the PRs... that'll teach me.
I guess that means I should do a proper review now.

**Comment by SergeGautherie** [comment]
Unbitrotted license header.

**Comment by SergeGautherie** [comment]
Could someone un-`WIP` and commit this PR, to ease related PR work?

**Comment by gigaherz** [comment]
But please, if it's not WIP, don't keep adding new commits. :P

**Comment by SergeGautherie** [comment]
> if it's not WIP, don't keep adding new commits

This PR is complete. Last non-trivial commit I made was a month ago, while the PR was stalled in WIP state.

File: modules/rostests/kmtests/kmtest/service.c [hunk 3]
```diff
@@ -101,6 +102,102 @@ KmtCreateService(
...
+            break;
+    }
+
+    if (ServiceStatus.dwCurrentState != ExpectedServiceState)
+    {
+        ok(FALSE, "Service = %ls, state = %lu %s (!= %lu %s)\n",
+           ServiceNameOut,
+           ServiceStatus.dwCurrentState, KmtGetServiceStateAsString(ServiceStatus.dwCurrentState),
+           ExpectedServiceState, KmtGetServiceStateAsString(ExpectedServiceState));
+        goto cleanup;
```
Comment (ThFabba) [inline modules/rostests/kmtests/kmtest/service.c]: Wouldn't you want to return failure in this case?

File: modules/rostests/kmtests/kmtest/service.c [hunk 4]
```diff
@@ -101,6 +102,102 @@ KmtCreateService(
...
+ * @param ServiceHandle
+ *        Handle to the service
+ * @param ExpectedServiceState
+ *        State which the service should be in
+ *
+ * @return Win32 error code
+ */
+static
+DWORD
+KmtCheckServiceState(
```
Comment (ThFabba) [inline modules/rostests/kmtests/kmtest/service.c]: This doesn't just check, it waits. Maybe "ensure" is a better term?

File: modules/rostests/kmtests/kmtest/service.c [hunk 5]
```diff
@@ -101,6 +102,102 @@ KmtCreateService(
...
+ * @name KmtGetServiceStateAsString
+ *
+ * @param ServiceState
+ *        Service state as a number
+ *
+ * @return Service state as a string
+ */
+static
+PCSTR
+KmtGetServiceStateAsString(
```
Comment (ThFabba) [inline modules/rostests/kmtests/kmtest/service.c]: This seems overkill for a trace message in an error case. I'd say people can look up the numbers if they come by that unexpected case.

File: modules/rostests/kmtests/kmtest/service.c [hunk 6]
```diff
@@ -101,6 +102,102 @@ KmtCreateService(
...
+
+    assert(ServiceHandle);
+    assert(ExpectedServiceState);
+
+    if (!QueryServiceStatus(ServiceHandle, &ServiceStatus))
+        error_goto(Error, cleanup);
+
+    while (ServiceStatus.dwCurrentState != ExpectedServiceState)
+    {
+        // NB: Not bothering with ServiceStatus.dwWaitHint nor ServiceStatus.dwCheckPoint.
```
Comment (ThFabba) [inline modules/rostests/kmtests/kmtest/service.c]: Not sure I understand this comment. "bothering" to do it is less code than the length of this comment.

File: modules/rostests/kmtests/kmtest/service.c (hunk 3 above)
Comment (SergeGautherie) [reply modules/rostests/kmtests/kmtest/service.c]: I could if you prefer, though my reasoning was: fail for subroutine unexpected error only.
Here, we succeed at checking what we want. Only the result is false, which we do report as a test failure.
In other word, it is a minimal change of behavior, that can later be made harder if need be.
This is also kind of why I initially called it "Check" only.

**Comment by SergeGautherie** [comment]
@ThFabba, this would really help:
- Easier testing, at least for me.
- Recent ROSTESTS-300 not going unnoticed.

**Pull request merged**